OpenRA.Game/Graphics/Util.cs

```diff
@@ -20,12 +20,26 @@ public static class Util
 		// yes, our channel order is nuts.
 		static readonly int[] ChannelMasks = { 2, 1, 0, 3 };
 
-		public static void FastCreateQuad(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv, in float3 size, in float3 tint, float alpha)
+		public static void FastCreateQuad(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv,
+										  in float3 size, in float3 tint, float alpha, in int rotAngle = 0)
 		{
+			var a = o;
 			var b = new float3(o.X + size.X, o.Y, o.Z);
 			var c = new float3(o.X + size.X, o.Y + size.Y, o.Z + size.Z);
 			var d = new float3(o.X, o.Y + size.Y, o.Z + size.Z);
-			FastCreateQuad(vertices, o, b, c, d, r, samplers, paletteTextureIndex, tint, alpha, nv);
+
+			// Rotate sprite if rotation angle is not equal to 0
+			if (rotAngle != 0)
+			{
+				var centerPoint = o + (size / 2);
+				var rotWAngle = new WAngle(rotAngle);
+				a = RotatePoint(o, centerPoint, -rotWAngle);
```

Comment:
`FastCreateQuad` is a hot path so we want to avoid unnecessary overhead.

This can be done much more efficiently by calculating the center point, calculating the rotated delta-x and delta-y once, and then adding the deltas in the different combinations to get each of the four corners.

Reply:
I have moved the Sin and Cos calculations out of `RotatePoint()` as they were being calculated multiple times when they only need to be calculated once. This leaves a few multiplications and additions within the `RotatePoint()` function, which I don't think can be brought out easily, as the deltas are not the same for each point. It may be possible with some significant rework, but my math is not quite at the level to figure this out. Still I think with the Sin and Cos being only calculated once, this version should be very close to optimal.